### PR TITLE
JA4 for TLS and QUIC -- v11

### PR DESCRIPTION
**Pull request by satta**
Previous PR: #10725

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/6379

Changes to previous PR:
* Rebase against current master.
* Put stray JA4 line into correct commit.
* Add comment explaining why `nof_exts` is used as a separate field.
* Use Rust's `std::cmp::min` instead of using explicit `if`s.
* Adopt the new style guard headers prefixed with `SURICATA` and without leading/trailing underscores.
* Condense JA4/JA3 documentation into single RST document.

```
SV_BRANCH=pr/1761
```

### CODE REVIEW

**Comment by satta** [comment]
No idea about the failing test, seems unrelated to the PR.

**Comment by victorjulien** [comment]
> No idea about the single failing test, seems unrelated to the PR.

Agreed, a restart fixed it.

File: src/app-layer-ssl.c [hunk 1]
```diff
@@ -834,37 +844,49 @@ static inline int TLSDecodeHSHelloCipherSuites(SSLState *ssl_state,
...
+                return -1;
+        }
 
         uint16_t processed_len = 0;
         /* coverity[tainted_data] */
         while (processed_len < cipher_suites_length)
         {
             if (!(HAS_SPACE(2))) {
-                Ja3BufferFree(&ja3_cipher_suites);
+                if (SC_ATOMIC_GET(ssl_config.enable_ja3)) {
```
Comment (victorjulien) [inline src/app-layer-ssl.c:861]: is there a race condition here? If `ssl_config.enable_ja3` flips to `false` after the alloc, but before this check, then we'd not free it? I wonder if it makes more sense to init it to a const
`const bool enable_ja3 = SC_ATOMIC_GET(ssl_config.enable_ja3);`

Comment (victorjulien) [reply src/app-layer-ssl.c:861]: alternatively, just check the pointer `ja3_cipher_suites`. If it is non-NULL, free it regardless?

Comment (victorjulien) [reply src/app-layer-ssl.c:861]: for the logic below, the ptr check could also work I think

**Comment by victorjulien** [comment]
Thanks @satta, I think this looks really good. A few comments inline.

File: src/app-layer-ssl.c (hunk 1 above)
Comment (satta) [reply src/app-layer-ssl.c:861]: `ssl_config.enable_ja3` is only set in `RegisterSSLParsers()` (which I assume is only called once during the initial setup) and `SSLEnableJA3()` which is only called during rule keyword setup (to enable the feature if a rule uses it). I would expect any of these calls to return before any parsing ever takes place.

But it doesn't hurt to store the value at the beginning of `TLSDecodeHSHelloCipherSuites()` in a local `const` value to ensure it definitely stays consistent across the whole runtime of the function. I'd be happy to make this change.

Comment (satta) [reply src/app-layer-ssl.c:861]: If that is all you have, I'd just push one more PR.

**Comment by satta** [comment]
New PR: #10836

**Pull request closed without merging**